sandbox/libs/authn/src/main/java/org/opensearch/authn/jwt/JwtVerifier.java

```diff
@@ -26,7 +26,7 @@
 public class JwtVerifier {
     private final static Logger log = LogManager.getLogger(JwtVerifier.class);
 
-    public static JwtToken getVerifiedJwtToken(String encodedJwt) throws BadCredentialsException {
+    public static JwtToken getVerifiedJwtToken(String encodedJwt) throws RuntimeException {
```

Comment:
Why the `RuntimeException` all over this class? I think these methods should continue to throw BadCredentialsException and get handled gracefully by the caller of these methods.

Reply:
Peter had said to extend RuntimeException instead of using BadCredentialsException all over the place. Maybe I misunderstood? This is the old PR which had the comment https://github.com/opensearch-project/OpenSearch/pull/5528

Reply:
I think @peternied 's comment is that the `login` method signature shouldn't change. It probably means that the exception was not being handled within an implementation of `login` somewhere so the IDE added it to the interface.

Reply:
Gotcha. I misunderstood that then. I will go ahead and correct these.

Reply:
I updated things let me know if this is more in-line with what you are thinking

Reply:
FYI the Shiro subject interface throws an AuthenticationException, a catchall for any exception that occurs during Authentication.

See: https://github.com/apache/shiro/blob/main/core/src/main/java/org/apache/shiro/subject/Subject.java#L332

As @peternied mentioned, if we do change the signature for login it could throw a general AuthenticationException.

FYI There are 2 high-level types of exceptions in Java.

- Checked exception
- Unchecked exception (RuntimeException is unchecked)

Unchecked exceptions, like RuntimeException, don't require modifying a method signature to indicate it throws that kind of exception. Image all of the methods out there that would need to be annotated with `throws NullPointerException` if it was required to throw and catch them. 

There's been discussion before on designing methods and pros and cons of different designs. I'll see if I can find the discussion.

Option 1:

```
interface Subject {
   boolean login();
}

...

// code called when logging in.
boolean authenticated = subject.login(authToken);
// if authenticated is false then respond with 401
```

Option 2:

```
interface Subject {
   void login() throws AuthenticationException;
}

...

// code called when logging in.
try {
   subject.login(authToken);
} catch (AuthenticationException ae) {
   // respond with 401
}
```

They are both valid ways of designs and the caller will have to use them differently.

Reply:
Thank you for sharing. I think in this case, I just was not sure what Peter meant by "off of" and went one way instead of the other.

As far as the login stuff is concerned, my understanding is that we do not have that level of control over the implementation. I want to say that if we want to use only unchecked exceptions we would need to use something other than BadCredentialsException as its currently implemented. I think this is because an exception is either checked or unchecked and cannot be case-by-case adjusted. We could overwrite it however or make our own which I believe Peter was alluding to in the previous PR--I could also be wrong though.

Reply:
I agree with the notion of RuntimeException being way too generic, and so BadCredentialsException should be used as it fits the scenario, or if it helps we can a create a JWT specific exception class, says `InvalidJWTException.java` and have that be thrown. This class can extend shiro's AuthenticationException if needed by design.